lockfile/src/spdx.rs

```diff
@@ -76,62 +79,84 @@ fn type_from_url(url: &str) -> anyhow::Result<PackageType> {
     }
 }
 
+fn from_purl(pkg_url: &str, pkg_info: &PackageInformation) -> anyhow::Result<Package> {
+    let purl = PackageUrl::from_str(pkg_url)?;
+
+    let package_type = PackageType::from_str(purl.ty())
+        .or_else(|_| type_from_url(&pkg_info.download_location))
+        .context(UnknownEcosystem)?;
+
+    let name = match (package_type, purl.namespace()) {
+        (PackageType::Maven, Some(ns)) => format!("{}:{}", ns, purl.name()),
+        (PackageType::Npm | PackageType::Golang, Some(ns)) => format!("{}/{}", ns, purl.name()),
+        _ => purl.name().into(),
+    };
+
+    let pkg_version = match (&pkg_info.version_info, purl.version()) {
+        (Some(v), _) => v.to_string(),
+        (None, Some(v)) => v.into(),
+        _ => bail!("Version not found for `{}`", pkg_info.name),
+    };
+
+    let version = purl
+        .qualifiers()
+        .iter()
+        .find_map(|(key, value)| match key.as_ref() {
+            "repository_url" => Some(PackageVersion::ThirdParty(ThirdPartyVersion {
+                version: pkg_version.clone(),
+                registry: value.to_string(),
+            })),
+            "download_url" => Some(PackageVersion::DownloadUrl(value.to_string())),
+            "vcs_url" => {
+                if value.as_ref().starts_with("git+") {
+                    Some(PackageVersion::Git(value.to_string()))
+                } else {
+                    None
+                }
+            },
+            _ => None,
+        })
+        .unwrap_or(PackageVersion::FirstParty(pkg_version));
+
+    Ok(Package { name, version, package_type })
+}
+
+fn from_locator(registry: &str, locator: &str) -> anyhow::Result<Package> {
+    lazy_static! {
+        static ref RE: Regex = Regex::new(r"^(([^:]*:)*[^@]+)@([^:]+)$").unwrap();
+    }
+
+    let package_type = PackageType::from_str(registry).map_err(|_| UnknownEcosystem)?;
+
+    let captures = RE.captures(locator).expect("Invalid locator");
+    let name = decode(captures.get(1).expect("Invalid package name").as_str())?;
+    let version = captures.get(3).expect("Invalid package name").as_str();
```

Comment:
```suggestion
    let version = captures.get(3).expect("Invalid package version").as_str();
```
This is a guess, based on the context.